### PR TITLE
bpo-38858: Allocate small integers on the heap

### PR DESCRIPTION
**Pull request by vstinner**
Allocate small Python integers (small_ints of longobject.c) on the
heap, rather than using static objects.

https://bugs.python.org/issue38858

### CODE REVIEW

File: Objects/longobject.c [hunk 1]
```diff
@@ -5784,40 +5783,30 @@ int
...
-            assert(Py_SIZE(op) == size);
-            assert(v->ob_digit[0] == (digit)abs(ival));
-        }
-        else {
-            (void)PyObject_INIT(v, &PyLong_Type);
+    for (Py_ssize_t i=0; i < NSMALLNEGINTS + NSMALLPOSINTS; i++) {
+        sdigit ival = (sdigit)i - NSMALLNEGINTS;
+        int size = (ival < 0) ? -1 : ((ival == 0) ? 0 : 1);
+
+        PyLongObject *v = _PyLong_New(1);
```
Comment (pablogsal) [inline Objects/longobject.c:5790]: This looks incorrect, shouldn't it be "i" instead of "1"?

Comment (vstinner) [reply Objects/longobject.c:5790]: The argument is the number of digits: small integers have always 1 digit, no?

**Comment by vstinner** [comment]
This change prepares the code to have small integers per interpreter (to support subinterpreters), rather than sharing these singletons between multiple interpreters.

File: Objects/longobject.c (hunk 1 above)
Comment (pablogsal) [reply Objects/longobject.c:5790]: Oh, I totally misread the function name, my bad :(

Comment (vstinner) [reply Objects/longobject.c:5790]: No problem, I also had to read _PyLong_New() code twice to understand how it is supposed to work :-D I am more used the public PyLong_FromLong().

**Pull request merged**